### ISSUE TITLE
SELECT … INTO new_table should update list of available tables

### ISSUE THREAD

**Issue by drocco007** [dbcli/pgcli#211]
In this screenshot, my_new_table does not appear in the list of table suggestions:

![select_into_no_refresh](https://cloud.githubusercontent.com/assets/1424858/7217263/d5cfc39c-e5ec-11e4-8178-b6a460e35267.png)

**Comment by amjith** [comment]
The auto-refresh of tables and columns are triggered by specific commands (https://github.com/amjith/pgcli/blob/master/pgcli/main.py#L391). 

I've never seen 'SELECT * INTO new_table FROM existing_table'. That's an awesome feature. :)

But having an auto-refresh for SELECT statement seems overkill, since that is the most common statement. 

I wonder if there is an easy way to detect if the metadata has changed and then issue an auto-refresh. 

For example, after each command, run a query that returns a "hash" or a "count" of available tables/views etc from the current database. Then we check if that number has changed from previous time and then issue an auto-refresh. 

Thoughts?

**Comment by darikg** [comment]
That's a cool idea -- it would also handle cases like `SELECT my_function_that_creates_a_new_table()`

**Comment by amjith** [comment]
So I just found this talk by someone named @drocco007. https://www.youtube.com/watch?v=HvMj8srOh-o

I'm wondering if it's possible to setup a NOTIFY channel for table changes. So anytime a new table is created or altered, can I be notified? 

If so, can we use that in pgcli to setup completion refresh?

**Comment by drocco007** [comment]
<grin>

The immediate hurdle I can see here is that normally you set up a trigger
that fires the notification that you want when something changes in the
database. We could probably set a trigger up to fire when the catalog
changes, but that feels a little weird to me, since we'd effectively be
messing with end user's databases. But maybe that's just me being overly
cautious. What do you folks think?

In the meantime, I'm going to see if I can figure out what psql does to
handle this.

**Comment by darikg** [comment]
Does that mean the user would have to have `create trigger` permissions?

**Comment by amjith** [comment]
Interesting point about user privileges to create triggers. I didn't even think about that. 

Also @drocco007 has a point about messing with user's notify triggers. It does make me nervous. So feel free to shoot down the suggestion, I just got excited when I saw the video, it just seemed like such an elegant solution. I didn't think it through fully (something about late night coding and lack of coffee).

We can go back to running an inexpensive query to detect if the number of tables have changed and trigger an auto-refresh based on that. It won't detect when a table schema is altered (say adding a new column), but that's a corner case I'm willing to live with.

**Comment by j-bennet** [comment]
I like the idea of an inexpensive query to find out table count. I think even finding table column counts may not end up to be a "heavy" query. Something like this:

select t.table_name, count(c.column_name) as column_count
from information_schema.tables t
join information_schema.columns c on c.table_name = t.table_name
where t.table_schema = 'public' group by t.table_name;

Especially if it's only run after an "ALTER" statement. Yes, it is possible to do "SELECT my_function_that_alters_a_table()", but I think this would be a very edgy edge case and maybe it's ok if we don't detect it.

**Comment by j-bennet** [comment]
Connected: https://github.com/dbcli/pgcli/issues/225

**Comment by darikg** [comment]
Here's a query that hashes the column metadata:

```
SELECT  md5(array_agg(
                    --schemaname, tablename, columnname
                ROW(nsp.nspname, cls.relname, att.attname)
                ORDER BY nsp.nspname, cls.relname, att.attname
            )::TEXT)
FROM    pg_catalog.pg_attribute att
        INNER JOIN pg_catalog.pg_class cls
            ON att.attrelid = cls.oid
        INNER JOIN pg_catalog.pg_namespace nsp
            ON cls.relnamespace = nsp.oid
WHERE   cls.relkind IN ('r', 'v', 'm')
        AND NOT att.attisdropped
        AND att.attnum  > 0
```

I think array_agg and ordered aggregate functions are a relatively new feature though

**Comment by drocco007** [comment]
I took a look at psql's source, which takes a different tack here: instead of trying to keep a list of completions up to date, psql obtains the data lazily, issuing the queries for the needed artifacts on demand. They don't even bother caching the results within a single command (e.g. multiple tables joined in a single SELECT), which can easily be seen using `strace` on Linux. If you'd like to try it for yourself (I found using two terminals to be less confusing), do this:

    $ strace -e sendto -s 2048 -o /tmp/psql_trace.log psql ## rest of psql args

and `tail -f /tmp/psql_trace.log` in the other terminal. In your psql shell, try doing a SELECT and Tab-completing table names; you can see the queries being issued in the strace log.

I copied one of the table name queries it issued by hand and ran it directly, which takes about 5 ms for a local psql process connected to a server in a Docker container. This seems like a viable approach to me, and would have the added benefit of handling some of the corner cases like table-generating functions. If you folks are interested, I'd be happy to try and put together a prototype implementing this for just table names, say, which would give us a feel for the complexity and performance without requiring a huge investment of effort. If we like the prototype, we could then proceed to implement some of the other completion checks the same way.

What do you folks think?

**Comment by amjith** [comment]
Go forth and take a stab. :)

Here are a few concerns: 

* The 5ms round trip is an ideal case (you already know this). When the database instance is remote, it could be much longer. So having some strict timeouts would be nice.
* Imagine a database with 100s or 1000s of tables. Running this query would send back payloads that are much bigger, which is a strain on the database instance. So all it takes is one user with pgcli to drastically reduce the performance of a database instance that could be serving tons of requests from a web app.

Take all these concerns with a grain of salt. Because, I've dismissed certain things as impossible and each one of you have proved me wrong in various occasions with creative solutions. So I've learned not to say no to things.
